Error using withStyles as decorator in TypeScript

### Problem description

The withStyles function requires the wrapped component type to have a required `classes` prop, but the returned decorated class has an optional `classes` prop. This causes the `setState` method signatures to differ, so TypeScript throws a compile error.

### Steps to reproduce

Minimal code example:

```typescript
interface ExampleClasses {
  root: string
}

interface ExampleProps {
  text: string
}

@withStyles<ExampleProps, ExampleClasses>({
  root: {
    color: 'blue'
  }
})
export default class Example extends Component<ExampleProps & { classes: ExampleClasses }, {}> {
  render (): JSX.Element | null | false {
    return <p className={this.props.classes.root}>{this.props.text}</p>
  }
}
```

Compiler error:

```
error TS1238: Unable to resolve signature of class decorator when called as an expression.
  Type 'ComponentClass<ExampleProps & StyledComponentProps<ExampleClasses>>' is not assignable to type 'typeof Example'.
    Type 'Component<ExampleProps & StyledComponentProps<ExampleClasses>, ComponentState>' is not assignable to type 'Example'.
      Types of property 'setState' are incompatible.
        Type '{ <K extends never>(f: (prevState: ComponentState, props: ExampleProps & StyledComponentProps<Exa...' is not assignable to type '{ <K extends never>(f: (prevState: {}, props: ExampleProps & { classes: ExampleClasses; }) => Pic...'.
          Types of parameters 'f' and 'f' are incompatible.
            Types of parameters 'props' and 'props' are incompatible.
              Type 'ExampleProps & StyledComponentProps<ExampleClasses>' is not assignable to type 'ExampleProps & { classes: ExampleClasses; }'.
                Type 'ExampleProps & StyledComponentProps<ExampleClasses>' is not assignable to type '{ classes: ExampleClasses; }'.
                  Types of property 'classes' are incompatible.
                    Type 'ExampleClasses | undefined' is not assignable to type 'ExampleClasses'.
                      Type 'undefined' is not assignable to type 'ExampleClasses'.
```

### Versions

- Material-UI: 1.0.0-beta.8
- React: 15.6.1
- types/react: 15.6.1
- TypeScript: 2.5.2

<!-- Have a QUESTION? Please ask in [StackOverflow or gitter](http://tr.im/77pVj before opening an issue.

If you are having an issue with click events, please re-read the [README](http://tr.im/410Fg) (you did read the README, right? :-) ).

If you think you have found a _new_ issue that hasn't already been reported or fixed in HEAD, please complete the template above.

For feature requests, please delete the template above and use this one instead:

### Description
### Images & references

-->

@marcusjwhelan JSS is the system Material UI 1.0 uses internally for scoped CSS classes.

Having a required `classes` prop for the inner component that will be wrapped seems to be the standard in Material UI:

https://github.com/callemall/material-ui/blob/v1-beta/src/Badge/Badge.js#L72

The `withStyles` HOC always injects the `classes` prop into the inner component, and then in the resulting wrapped component, `classes` is exposed as an *optional* prop for overrides. The inner component expects classes to be provided by its parent, which is almost always going to be the wrapping component provided by `withStyles`. In your code, I'm still not seeing an access of `this.props.classes.whatever` which is what is confusing me. Where do you actually access the classes provided by the `withStyles` HOC?

Anyway, this will still not work as a decorator, because TypeScript enforces that class decorators return a type that is compatible with the original decorated class. The `withStyles` HOC changes the `classes` prop from required to optional, which means the types of some component methods are changed and the classes are not compatible. See the following definitions:


https://github.com/callemall/material-ui/blob/v1-beta/src/styles/withStyles.d.ts#L27


https://github.com/callemall/material-ui/blob/v1-beta/src/index.d.ts#L12-L16

Furthermore, if the inner component does not require a `classes` prop and `withStyles` is used as a decorator in Material UI version 1.0.0-beta.8, this error results:

```
Argument of type 'typeof Example' is not assignable to parameter of type 'ComponentType<ExampleProps & { classes: ExampleClasses; theme?: Theme<{}> | undefined; }>'.
  Type 'typeof Example' is not assignable to type 'StatelessComponent<ExampleProps & { classes: ExampleClasses; theme?: Theme<{}> | undefined; }>'.
    Type 'typeof Example' provides no match for the signature '(props: ExampleProps & { classes: ExampleClasses; theme?: Theme<{}> | undefined; } & { children?: ReactNode; }, context?: any): ReactElement<any> | null'.
src/ui/navigation/ButtonNav/LinkButton.tsx(23,37): error TS2339: Property 'classes' does not exist on type 'Readonly<{ children?: ReactNode; }> & Readonly<ExampleProps>'.
src/ui/navigation/ButtonNav/__tests__/LinkButton.test.tsx(34,9): error TS2339: Property 'href' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Example> & Readonly<{ children?: ReactNode; }> & R...'.
src/ui/navigation/ButtonNav/__tests__/LinkButton.test.tsx(43,9): error TS2339: Property 'href' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Example> & Readonly<{ children?: ReactNode; }> & R...'.
```

--- a/src/styles/withStyles.d.ts
+++ b/src/styles/withStyles.d.ts
@@ -20,9 +20,22 @@ export interface WithStylesOptions {
   name?: string;
 }
 
-export default function withStyles<P = {}, ClassNames = {}>(
+declare function withStyles(
+  style: StyleRules | StyleRulesCallback,
+  options?: WithStylesOptions
+): <
+  C extends React.ComponentType<P & { classes: ClassNames; theme?: Theme }>,
+  P = {},
+  ClassNames = {}
+>(
+  component: C
+) => C & React.ComponentClass<P & StyledComponentProps<ClassNames>>
+
+declare function withStyles<P = {}, ClassNames = {}>(
   style: StyleRules | StyleRulesCallback,
   options?: WithStylesOptions
 ): (
   component: React.ComponentType<P & { classes: ClassNames; theme?: Theme }>
-) => React.ComponentClass<P & StyledComponentProps<ClassNames>>;
+) => React.ComponentClass<P & StyledComponentProps<ClassNames>>
+
+export default withStyles;

--- a/test/typescript/styles.spec.tsx
+++ b/test/typescript/styles.spec.tsx
@@ -27,7 +27,10 @@ interface StyledComponentProps {
 
 const Component: React.SFC<
   StyledComponentProps & { classes: StyledComponentClassNames }
-> = ({ classes, text }) => <div className={classes.root}>{text}</div>;
+> = ({ classes, text }) =>
+  <div className={classes.root}>
+    {text}
+  </div>;
 
 const StyledComponent = withStyles<
   StyledComponentProps,
@@ -77,16 +80,19 @@ const customTheme = createMuiTheme({
 function OverridesTheme() {
   return (
     <MuiThemeProvider theme={theme}>
-      <Button>{'Overrides'}</Button>
+      <Button>
+        {'Overrides'}
+      </Button>
     </MuiThemeProvider>
   );
 }
 
 // withTheme
 
-const ThemedComponent: React.SFC<{ theme: Theme }> = ({ theme }) => (
-  <div>{theme.spacing.unit}</div>
-);
+const ThemedComponent: React.SFC<{ theme: Theme }> = ({ theme }) =>
+  <div>
+    {theme.spacing.unit}
+  </div>;
 const ComponentWithTheme = withTheme(ThemedComponent);
 
 // withStyles + withTheme
@@ -95,10 +101,25 @@ interface AllTheProps {
   classes: StyledComponentClassNames;
 }
 
-const AllTheStyles: React.SFC<AllTheProps> = ({ theme, classes }) => (
-  <div className={classes.root}>{theme.palette.text.primary}</div>
-);
+const AllTheStyles: React.SFC<AllTheProps> = ({ theme, classes }) =>
+  <div className={classes.root}>
+    {theme.palette.text.primary}
+  </div>;
 
 const AllTheComposition = withTheme(
   withStyles<{ theme: Theme }, StyledComponentClassNames>(styles)(AllTheStyles)
 );
+
+@withStyles(styles)
+class DecoratedComponent extends React.Component<
+  StyledComponentProps & { classes: StyledComponentClassNames }
+> {
+  render() {
+    const { classes, text } = this.props;
+    return (
+      <div className={classes.root}>
+        {text}
+      </div>
+    );
+  }
+}

--- a/test/typescript/styles.spec.tsx
+++ b/test/typescript/styles.spec.tsx
@@ -110,6 +110,7 @@ const AllTheComposition = withTheme(
   withStyles<{ theme: Theme }, StyledComponentClassNames>(styles)(AllTheStyles)
 );
 
+// As decorator
 @withStyles(styles)
 class DecoratedComponent extends React.Component<
   StyledComponentProps & { classes: StyledComponentClassNames }

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -8,7 +8,8 @@
     "moduleResolution": "node",
     "forceConsistentCasingInFileNames": true,
     "suppressImplicitAnyIndexErrors": true,
-    "noEmit": true
+    "noEmit": true,
+    "experimentalDecorators": true
   },
   "include": ["src/**/*", "test/typescript/*"]
 }

